src/main/java/org/jenkinsci/plugins/workflow/job/console/NewNodeConsoleNote.java

```diff
@@ -0,0 +1,131 @@
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Console line with note printed when a new {@link FlowNode} is added to the graph.
+ * Defines the {@code pipeline-new-node} CSS class and several attributes which may be used to control subsequent behavior:
+ * <ul>
+ * <li>{@code nodeId} for {@link FlowNode#getId}
+ * <li>{@code startId} {@link FlowNode#getId} for {@link BlockStartNode}, else {@link BlockEndNode#getStartNode}, else absent
+ * <li>{@code enclosingId} the immediately enclosing {@link BlockStartNode}, if any
+ * <li>{@code label} for {@link LabelAction} if present
+ * </ul>
+ * @see AnnotatedLogAction#annotateHtml
+ */
+@Restricted(NoExternalUse.class)
+public class NewNodeConsoleNote extends ConsoleNote<WorkflowRun> {
+
+    /**
+     * Prefix used in metadata lines.
+     */
+    private static final String CONSOLE_NOTE_PREFIX = "[Pipeline] ";
+
+    public static void print(FlowNode node, TaskListener listener) {
+        try {
+            listener.annotate(new NewNodeConsoleNote(node));
+        } catch (IOException x) {
+            // never mind
+        }
+        listener.getLogger().println(CONSOLE_NOTE_PREFIX + node.getDisplayFunctionName());
+    }
+
+    private final @Nonnull String id;
+    private final @CheckForNull String enclosing;
+    private final @CheckForNull String start;
+
+    private NewNodeConsoleNote(FlowNode node) {
+        id = node.getId();
+        if (node instanceof BlockEndNode) {
+            enclosing = null;
+            start = ((BlockEndNode) node).getStartNode().getId();
+        } else {
+            Filterator<FlowNode> it = FlowScanningUtils.fetchEnclosingBlocks(node).filter(Predicates.not(Predicates.equalTo(node)));
+            enclosing = it.hasNext() ? it.next().getId() : null;
+            start = node instanceof BlockStartNode ? node.getId() : null;
+        }
+    }
+
+    @Override
+    public ConsoleAnnotator<?> annotate(WorkflowRun context, MarkupText text, int charPos) {
+        StringBuilder startTag = new StringBuilder("<span class=\"pipeline-new-node\" nodeId=\"").append(id);
+        if (start != null) {
+            startTag.append("\" startId=\"").append(start);
+        }
+        if (enclosing != null) {
+            startTag.append("\" enclosingId=\"").append(enclosing);
+        }
+        FlowExecution execution = context.getExecution();
+        if (execution != null) {
+            try {
+                FlowNode node = execution.getNode(id);
+                if (node != null) {
+                    LabelAction a = node.getAction(LabelAction.class);
+                    if (a != null) {
+                        String displayName = a.getDisplayName();
+                        assert displayName != null;
+                        startTag.append("\" label=\"").append(Util.escape(displayName)); // TODO is there some better way to escape for attribute values?
```

Comment:
Noting here for future review/integration that this appears to be how we propose to deal with rendering of branch labels -- however it requires rich rendering of annotation content by the UI to be useful (i.e. not visible in plaintext logs). 

Is that the right decision?   Not sure, the mechanism seems more straightforward in some ways.

Reply:
Right, this keeps the HTML leaner and does more of the fanciness on the client side.